src/gui/plugins/origin_axes/OriginAxesConfig.cc

```diff
@@ -0,0 +1,213 @@
+
+/////////////////////////////////////////////////
+void OriginAxesConfig::LoadConfig(const tinyxml2::XMLElement *)
+{
+  if (this->title.empty())
+    this->title = "Origin Axis";
+
+  ignition::gui::App()->findChild<
+      ignition::gui::MainWindow *>()->installEventFilter(this);
+}
+
+/////////////////////////////////////////////////
+bool OriginAxesConfig::eventFilter(QObject *_obj, QEvent *_event)
+{
+  if (_event->type() == ignition::gazebo::gui::events::Render::Type)
+  {
+    // This event is called in Scene3d's RenderThread, so it's safe to make
+    // rendering calls here
+    this->UpdateOriginArrows();
+  }
+
+  // Standard event processing
+  return QObject::eventFilter(_obj, _event);
+}
+
+/////////////////////////////////////////////////
+void OriginAxesConfig::UpdateOriginArrows()
+{
+  // Load grid if it doesn't already exist
+  if (!this->dataPtr->x_axis || !this->dataPtr->y_axis || !this->dataPtr->z_axis)
+    this->LoadOriginAxes();
+
+  // If grid was not loaded successfully, don't update
+  if (!this->dataPtr->x_axis || !this->dataPtr->y_axis || !this->dataPtr->z_axis)
+    return;
+
+  if (!this->dataPtr->dirty)
+    return;
+
+  this->dataPtr->dirty = false;
+
+  auto visual_x = this->dataPtr->x_axis;
+  if (visual_x)
+  {
+    visual_x->SetVisible(this->dataPtr->visible);
+  }
+
+  auto visual_y = this->dataPtr->y_axis;
+  if (visual_y)
+  {
+    visual_y->SetVisible(this->dataPtr->visible);
+  }
+
+  auto visual_z = this->dataPtr->z_axis;
+  if (visual_z)
+  {
+    visual_z->SetVisible(this->dataPtr->visible);
+  }
+}
+
+/////////////////////////////////////////////////
+void OriginAxesConfig::LoadOriginAxes()
+{
+  auto loadedEngNames = rendering::loadedEngines();
+  if (loadedEngNames.empty())
+    return;
+
+  // assume there is only one engine loaded
+  auto engineName = loadedEngNames[0];
+  if (loadedEngNames.size() > 1)
+  {
+    igndbg << "More than one engine is available. "
+      << "Grid config plugin will use engine ["
+        << engineName << "]" << std::endl;
+  }
+  auto engine = rendering::engine(engineName);
+  if (!engine)
+  {
+    ignerr << "Internal error: failed to load engine [" << engineName
+      << "]. Grid plugin won't work." << std::endl;
+    return;
+  }
+
+  if (engine->SceneCount() == 0)
+    return;
+
+  // assume there is only one scene
+  // load scene
+  auto scene = engine->SceneByIndex(0);
+  if (!scene)
+  {
+    ignerr << "Internal error: scene is null." << std::endl;
+    return;
+  }
+
+  if (!scene->IsInitialized() || scene->VisualCount() == 0)
+  {
+    return;
+  }
+
+  if (this->dataPtr->x_axis || this->dataPtr->y_axis || this->dataPtr->z_axis)
+    return;
+
+  // Create grid
+  ignwarn << "Creating origin axes" << std::endl;
+
+  auto root = scene->RootVisual();
+  this->dataPtr->x_axis = scene->CreateArrowVisual();
+  this->dataPtr->y_axis = scene->CreateArrowVisual();
+  this->dataPtr->z_axis = scene->CreateArrowVisual();
```

Comment:
On Gazebo-classic the origin axes were implemented using dynamic lines instead of arrows. Most of the time users need the origin to be long and without arrows, see [this comment](https://osrf-migration.github.io/gazebo-gh-pages/#!/osrf/gazebo/pull-requests/1700/page/1#comment-7000228). @iche033 can correct me if I'm wrong, but I believe the lines should be less expensive to render, since we don't have the arrow head and the extra triangles from the cylinders. I'm not advocating for a change, I just think it should be pointed out.

---

Having said all that, I think it's totally valid to add support for custom axes like [RViz's axes displays](http://wiki.ros.org/rviz/DisplayTypes/Axes), which ties into the work @Sarath18 will be doing :smile: To that end, here are some suggestions to make the current plugin more flexible:

* Add an option to set the pose
* Add an option to set the radius
* Add an option to change the arrows into lines

This is definitely feature creep for this PR, but I'll just mention it as an idea for the future:

* Let the user attach an axis to any entity that has a pose component

Reply:
yes lines are less expensive to render. I think a generic axes plugin would be useful, e.g. for visualizing frames. That could be done by changing the name of this plugin to just AxesPlugin and adding a pose widget.